RoaringBitmap/src/main/java/org/roaringbitmap/FastAggregation.java

```diff
@@ -262,12 +264,76 @@ public static RoaringBitmap naive_and(RoaringBitmap... bitmaps) {
       return new RoaringBitmap();
     }
     RoaringBitmap answer = bitmaps[0].clone();
-    for (int k = 1; k < bitmaps.length; ++k) {
+    for (int k = 1; k < bitmaps.length && !answer.isEmpty(); ++k) {
       answer.and(bitmaps[k]);
     }
     return answer;
   }
 
+  /**
+   * Computes the intersection by first intersecting the keys, avoids
+   * materialising containers.
+   *
+   * @param bitmaps the inputs
+   * @return the intersection of the bitmaps
+   */
+  public static RoaringBitmap workShyAnd(RoaringBitmap... bitmaps) {
+    long[] words = new long[1024];
+    RoaringBitmap first = bitmaps[0];
+    for (int i = 0; i < first.highLowContainer.size; ++i) {
+      char key = first.highLowContainer.keys[i];
+      words[key >>> 6] |= 1L << key;
+    }
+    int numContainers = first.highLowContainer.size;
+    for (int i = 1; i < bitmaps.length && numContainers > 0; ++i) {
+      numContainers = Util.intersectArrayIntoBitmap(words,
+              bitmaps[i].highLowContainer.keys, bitmaps[i].highLowContainer.size);
+    }
+    if (numContainers == 0) {
+      return new RoaringBitmap();
+    }
+    char[] keys = new char[numContainers];
+    int base = 0;
+    int pos = 0;
+    for (long word : words) {
+      while (word != 0L) {
+        keys[pos++] = (char)(base + Long.numberOfTrailingZeros(word));
+        word &= (word - 1);
+      }
+      base += 64;
+    }
+    Container[][] containers = new Container[numContainers][bitmaps.length];
```

Comment:
whats the max size of numContainers

Reply:
@kishoreg The value of numContainers is smaller or equal to the smallest number of container from any one of the bitmaps.

Reply:
... but it is not trivial...

Reply:
You are basically trading memory for a binary search. That's probably not a win. See my other PR.

Reply:
> The value of numContainers is smaller or equal to the smallest number of container from any one of the bitmaps.

But in the previous loop, it is picking the first non-zero numContainer. Shouldn't it pick the min among all then?

Reply:
numContainers is the number of containers which can intersect and is bounded above by the number of containers in the smallest bitmap.

Reply:
numContainers decreases monotonically as the loop index increases, hence the last value is the smallest value.